### ISSUE TITLE
Write a grown-up widget library

### ISSUE THREAD

**Issue by Bloke** [textpattern/textpattern#968]
It's high time to phase out most/all of `txplib_forms.php` and parts of `txplib_html.php`. To achieve this, we should build an extensible OO widget library using inheritance and interfaces.

Benefits:
* Removal of the stupid overloaded parameter function signatures (like `fInput()` with 12 params!) in favour of clean constructors with setters.
* Cascading and compound widgets can be built more easily from the base components - in core and in plugins - reducing code.
* Potential to construct one or more tags that output widgets, which allows forms to be built.
* Potential to use the tags to build the admin-side UI from templates, which offers endless workflow customization options or easily buildable dashboards for people that know what they're doing.

Downsides:
* It's OO, so possibly slower than raw function calls. Needs to be optimised to maintain speed.
* Bigger product footprint overall because each discrete file has preamble text.

**Comment by bloatware** [comment]
+1, sure, please start with `upload_form(:-)`. The speed is not critical on the admin side, imo, and the extra footprint is justified if it helps to avoid code redundancy.

**Comment by philwareham** [comment]
I'm in favour of this massively. Maybe out could all justify the name Textpattern 5!? If we coincided it with unlimited custom fields of various types too that seems enough to warrant a new numbered release.

**Comment by philwareham** [comment]
Would you be considering adopting a PHP framework as part of this?

**Comment by bloatware** [comment]
I would rather investigate on the side of template languages (like mustache), to be able to process forms both server and client side.

**Comment by philwareham** [comment]
Sure. As long as the basic structure of HTML isn't a million miles away from what we currently have and is easy to style, then it gets my vote.

**Comment by Bloke** [comment]
Processing client and server forms did occur to me, but I never got my head round mustache fully. Seemed useful though. I've used Twig, fwiw but found the control structures lacking.

What I'm proposing here - at least initially - is just a way to maintain our notion of "checkbox", "radio set", "select input", "textbox" and so forth, along with their granular counterpart "tag", and their combined versions such as "file input", "checkbox set", "upload form", etc and just wrap them in a better, extensible manner. Like tag lego. We already have most of the components in place, it's just that they're:

a) functions
b) overloaded with parameters
c) not very flexible to change (requires introduction of new params)

So I was planning on simply writing a few quick objects to replicate the existing functionality in a more OO style, and make it easier to customize (and shorter to invoke than counting which positional argument to use when you need to specify an `id`!)

Longer term, the objects could be invoked from a tag (e.g. `<txp:widget type="upload" name="file" class="file-upload" multiple required />`) and you'd get a form field spat out. Value validation might be part of all this too (patterns), along with checking you've supplied 'mandatory' attributes (such as `name`) so things don't break.

Things this is NOT:

* A raw HTML building framework (although technically the Tag object could be used to build any tag or structure, it'd drive you mad to do so).
* A templating language, as such. It's just a series of shortcut objects for outputting common form controls or building your own, from smaller components (such as tag, checkbox, label, etc).
* Bloated. It should adopt sensible defaults so that most widgets require minimal customization, but the option to do so should be available.
* A breaking change. It's a new parallel system that core could adopt over time to render admin-side widgets, plugins and admin themes could use, and eventually, will be a replacement for `txplib_forms.php`.

Yes, I know, Not-Invented-Here. I've tried countless widget builders over the years and every one has either left me cold, been too limited, too big, or too complex trying to do everything. This isn't one of those systems.

I've made a start in a separate local branch, but my OO is flaky so it'll take a few iterations to figure out what extends what and interfaces what else. If the scope needs to expand to do more, and it makes sense to do so, let's hear proposals.

**Comment by bloatware** [comment]
> I was planning on simply writing a few quick objects to replicate the existing functionality in a more OO style, and make it easier to customize

That will be already a huge improvement, and if, one day, we are able to compose a whole admin pane from txp tags, my 10-years dream would come true.

Templating is probably off-topic here, but let me explain what I mean by example. Suppose we want to allow theme authors to construct their own upload image previews, with few extra fields (name, resize, thumbnails dimensions). Some of these fields (resize) could be pre-filled on the server side, from user preferences or last used values. Other fields (name) are only available to js. So we need a template language (even our own) that can be parsed both by php and js, and easy to use for theme authors. Something in the vein of
```
<div>
    <input name='title' value='{title}' />
    <span>{type} ({size} B)</span>
</div>
```

**Comment by Bloke** [comment]
> That will be already a huge improvement, and if, one day, we are able to compose a whole admin pane from txp tags, my 10-years dream would come true.

Yours and mine too :-)

Re: templating, we should definitely build in some kind of support for this. I don't know how yet. At the moment I have the following Objects:

* Tag - the basic thing that replaces `tag()`, `tag_start()`, `tag_end()`, and `tag_void()`. Create one of these, `setAtts()` on it as an array of name-value pairs with optional flags (that can be passed to `join_atts()`) adn call `render()` with an option of outputting the full tag with content, a self-closer, a start tag, an end tag, or verbatim if nothign matches. If you omit the option (for example, if you just echo the object via its `__toString()` magic method) then the most appropriate option is used. If you used `setContent()`, then you get a full tag with content, for example.
* Attribute - one or more attributes in the tag. Attributes can be grouped by properties (such as flags) so that when they are joined via `join_atts()` the appropriate flags are associated with the group. This permits things like in `fInput()` where all atts except `value` are governed by `TEXTPATTERN_STRIP_EMPTY`. Haven't decided if an Attribute should be just one thing and we have an AttributeCollection for many. My head says a collection is the 'right' thing to do, but my heart and performance practicalities are leaning towards a single object that contains an array of attributes.
* WidgetInterface - defines a few core methods that the tags must define. Not sure I've got this bit right.
* (a test object) Textbox - to replicate `fInput()` functionality. But does it extend `Tag`? or is `Tag` the most specific component and the other objects that are made up of the smaller parts the main players?

Thoughts and ideas welcome.

**Comment by Bloke** [comment]
Story so far, via examples:

## Textboxes

Render a bog standard input box with the given name and value. Note you can just `echo` or `print` the tag to output it directly.
```php
$tag = Txp::get('\Textpattern\Widget\Textbox', 'fish', 'cod');

print $tag;

Renders:
<input type="text" name="fish" value="cod" />
```

You can supply attributes via the `setAtts()` method, and the `value` param is optional. The `setBool()` method is a shortcut for setting one or more boolean attributes:
```php
$tag = Txp::get('\Textpattern\Widget\Textbox', 'valueless')
->setAtts(array('data-whatever' => 'my-custom-attribute'))
->setAtts(array('i-am-boolen' => true))
->setBool(array(
   'required',
   'some-future-boolean-option'
))
->setBool('autocomplete');

print $tag;

Renders:
<input type="text" name="valueless" data-whatever="my-custom-attribute" value="" i-am-boolen required some-future-boolean-option autocomplete />
```

## DIY tags
If you want to delve off-piste, you can go to the lowest level with the `Tag` class and build your own stuff:

```php
$label = Txp::get('\Textpattern\Widget\Tag', 'label')
->setAtts(array('for' => 'terms'))
->setContent('I agree to the terms');

$tag = Txp::get('\Textpattern\Widget\Tag', 'input')
->setAtts(array(
   'type' => 'checkbox',
   'class' => 't-and-c',
   'id' => 'terms',
   'value' => 1));

echo $tag . $label;

Renders:
<input type="checkbox" class="t-and-c" id="terms" value="1" />
<label for="terms">I agree to the terms</label>
```

## Separate start / end
Custom content? No problem. Here's a form with two fields inside it. Note that during rendering - using the `render()` method directly instead of the magic methods - we tell it to render the `start` of the form only, then the two input widgets, then the `end` of the form.

Also, the `action` attribute would normally be swallowed by the default 'strip empty' behaviour so we've modifed the flag for that attribute to permit empties. The same thing happened in the `Textbox` above for the `value` attribute by default, to permit the valueless attribute.

```php
$text = Txp::get('\Textpattern\Widget\Textbox', 'email')
->setAtts(array('type' => 'email'));

$area = Txp::get('\Textpattern\Widget\Tag', 'textarea')
->setAtts(array('placeholder' => "Tell me something I don't know"));

$form = Txp::get('\Textpattern\Widget\Tag', 'form')
->setAtts(array('name' => 'sendy', 'id' => 'my-form'))
->setAtts(
   array('action' => ''),
   array('flag' => TEXTPATTERN_STRIP_NONE)
);

echo
   $form->render('start'),
   $text->render(),
   $area->render('full'),
   $form->render('end');

Renders:
<form name="sendy" id="my-form" action="">
<input type="email" name="email" value="" />
<textarea placeholder="Tell me something I don&#039;t know"></textarea>
</form>
```

Sky's the limit.

Next up: a collection of objects that act as convenience wrappers around the `Tag` so you don't have to specify so many attributes. Hello checkbox, radioset, textarea, selects, ...

Questions or comments / improvements on implementation welcome.

**Comment by Bloke** [comment]
5c06fea9 introduces the `TagCollection` class which, as its name suggests, holds a bunch of tags. This is primarily of use for creating groups of related tags, and the example implementation bundled with the commit is for `RadioSet`. Essentially, it's an array that holds a set of widgets, in the order you add them to the object. So in the case of RadioSet if you give it an array of 3 `key=>label` pairs, it'll create:

* Radio 1
* Label 1
* Radio 2
* Label 2
* Radio 3
* Label 3

And when you print/render it, they're just concatenated in that order.

Below is an example of it in use. It's not much different to its functional counterpart. Note that there is no (direct) implementation of setting the `id`, as it's handled internally via concatenation of the `key` and `value`. In the old function-based version, there was an optional `id` parameter which was added as a prefix to the key-value pair. I'm not sure if there's an actual use case for this: radio sets need to have a unique name, so why prepend an extra value? Anyone any idea whether a manually-set ID prefix is of any practical purpose here?

Note also that the 'no' value is checked by setting the third parameter `$default` during construction. We should maybe add the ability to set the default value via a method too (e.g. `setDefault('some-value')`), though I can't think of a reason why it'd be useful right now.

```php
$radset = array(
   'yes' => gTxt('yes'),
   'no' => gTxt('no'),
   'maybe' => gTxt('maybe'),
);

$radioset =Txp::get('\Textpattern\Widget\RadioSet', 'decisionRadio', $radset, 'no');
echo $radioset;

Renders:
<input type="radio" name="decisionRadio" id="decisionRadio-yes" class="radio" value="yes" />
<label for="decisionRadio-yes">Yes</label>
<input checked type="radio" name="decisionRadio" id="decisionRadio-no" class="radio active" value="no" />
<label for="decisionRadio-no">No</label>
<input type="radio" name="decisionRadio" id="decisionRadio-maybe" class="radio" value="maybe" />
<label for="decisionRadio-maybe">maybe</label>
```

The convention is (as before) that if you set a value as default (checked), it gets the class `active` appended to it, and the `class` name itself is named after the widget (`radio` in this case). If you don't like that, you can build your own collection. Or, after building the stock collection, you can use the `getKeys()` and `getWidget()` methods to access the individual widgets and alter their properties by hand prior to rendering. We could probably add an Iterator to the Collection for even greater flexibility so you could use `foreach()` to step through its members.

The TagCollection concept should be extensible to select lists and checkbox sets, as well as widget combos. I'm hoping this means we can introduce a  sensible InputLabel as a convenience method, which will combine the given input widget with a label, and be able to raise a callback as we do now.

I've also introduced a dedicated `Label` convenience tag, which takes the `for` attribute as its second argument. It's used internally by `RadioSet` but you could create arbitrary label tags, with or without the `for`. The main advantage over the old way of doing things is, because all tags are based on the `Tag` object, you can easily `addAtts()` for any of them to provide custom attributes, without having to overload constructors with huge parameter chains. Keeps things nice and simple, with constructors only providing the bare functionality that you can then build on as needed.

Please take it for a spin and suggest improvements or add to the library if you wish.

**Comment by philwareham** [comment]
Just a minor thing: can any widgets that are generated have their attributes in the following order at the start of a HTML tag:

1. `class`
2. `id`
3. `name`
4. `type`

That makes it quicker for me to theme things, as the hooks I need are at the front of the tag so I can view them quickly. Basically, `class` should be the very first attribute on a tag as that is generally the one we hook to. Cheers!

**Comment by Bloke** [comment]
Now _that_ I don't know about quite yet!

I did spot yesterday that it seems to be putting them in a random order - albeit seemingly unchanging after the tag has been invoked - and I'm not sure why `join_atts()` is doing that. It seems to partially depend on how the object itself is created and which attributes are added first. It seems that booleans always go last. But if I `addAtts()` a data attribute right at the end, it appears first in the list - but always after `name` which seems to trump everything.

From what I can tell so far:

* added attributes appear in the order they are added.
* booleans always occur last, in the order they are added.
* name always appears first.

It's on my list of things to investigate, unless someone else wants to step in and hazard a guess why it's doing this.

**Comment by Bloke** [comment]
Attribute order should be fixed in be6559e9.

**Comment by jools-r** [comment]
I like the way this is going!!

**Comment by Bloke** [comment]
810cf759 adds select list support. Examples:

``` php
$optset = array(
   'banana' => 'Banana',
   'strawberry' => 'Strawberry',
   'vanilla' => 'Vanilla',
   'chocolate' => 'Chocolate',
);

$select = Txp::get('\Textpattern\Widget\Select', 'flavour_love', $optset, 'banana, chocolate');

echo $select->render();

Renders:
<select name="flavour_love[]" multiple>
<option value="banana" selected dir="auto">Banana</option>
<option value="strawberry" dir="auto">Strawberry</option>
<option value="vanilla" dir="auto">Vanilla</option>
<option value="chocolate" selected dir="auto">Chocolate</option>
</select>
```
Note that because you supplied more than one default value, the select was automatically converted to a `multiple` select. You can always use `setMultiple()` or build the select manually if you prefer: 
``` php
$select2 = Txp::get('\Textpattern\Widget\Select', 'flavour_love')
->addOption('banana', 'Banana')
->addOption('pineapple', 'Pineapple')
->addOption('strawberry', 'Strawberry', true)
->addOption('chocolate', 'Chocolate')
->setBool('required')
->setAtts(array('class' => 'selectClass', 'data-whatnot' => 'some value'))
->blankFirst();

echo $select2->render();

Renders:
<select name="flavour_love" class="selectClass" data-whatnot="some value" required>
<option value="">&#160;</option>
<option value="banana" dir="auto">Banana</option>
<option value="pineapple" dir="auto">Pineapple</option>
<option value="strawberry" selected dir="auto">Strawberry</option>
<option value="chocolate" dir="auto">Chocolate</option>
</select>
```

**Comment by Bloke** [comment]
I should mention 1e884730 here, which introduces something we don't currently have in our arsenal: OptGroups. Here's an example if you want to directly build a select list:
``` php
$mcu = Txp::get('\Textpattern\Widget\Select', 'marvel_characters')
->addOptGroup('Heroes')
->addOption('spiderman', 'Spiderman')
->addOption('iron-man', 'Iron Man', true)
->addOption('thor', 'Thor')
->addOption('professor-x', 'Professor X')
->addOptGroup('Villains')
->addOption('loki', 'Loki')
->addOption('magneto', 'Magneto')
->addOption('ultron', 'Ultron')
->addOption('doctor-doom', 'Doctor Doom')
->setBool('required')
->setAtts(array('class' => 'marvelClass', 'data-mcu' => 'Marvel Cinematic Universe'))
->addOption('black-widow', 'Black Widow', false, 'Heroes');
echo $mcu;
```
Renders:
``` html
<select name="marvel_characters" class="marvelClass" data-mcu="Marvel Cinematic Universe" required>
<optgroup label="Heroes" dir="auto">
<option value="spiderman" dir="auto">Spiderman</option>
<option value="iron-man" selected dir="auto">Iron Man</option>
<option value="thor" dir="auto">Thor</option>
<option value="professor-x" dir="auto">Professor X</option>
<option value="black-widow" dir="auto">Black Widow</option>
</optgroup>
<optgroup label="Villains" dir="auto">
<option value="loki" dir="auto">Loki</option>
<option value="magneto" dir="auto">Magneto</option>
<option value="ultron" dir="auto">Ultron</option>
<option value="doctor-doom" dir="auto">Doctor Doom</option>
</optgroup>
</select>
```
Note we accidentally forgot Black Widow, and added her afterwards by specifying the group to add her to. Under normal operation, when it encounters an `addOptgroup()` it sets it as a 'sticky' group so anything that occurs after it appears in the that group. This permits linear construction of option groups. The extra parameter allows you to go back and add stuff to a specific group at a later date.

Alternatively, you can build your groups separately and add them as content to a select. This gives you more control over each group as you can set attributes directly on each, disable groups, and so forth. Here we add a `class` and `data-` attribute to each group:
``` php
$mcu2 = Txp::get('\Textpattern\Widget\Select', 'marvel_characters');
$optGroup1 = Txp::get('\Textpattern\Widget\OptGroup', 'Heroes')
->addOption('spiderman', 'Spiderman')
->addOption('iron-man', 'Iron Man', true)
->addOption('thor', 'Thor')
->addOption('professor-x', 'Professor X')
->addOption('black-widow', 'Black Widow')
->setAtts(array('class' => 'marvelClass', 'data-mcu' => 'Marvel Heroes'));
$optGroup2 = Txp::get('\Textpattern\Widget\OptGroup', 'Villains')
->addOption('loki', 'Loki')
->addOption('magneto', 'Magneto')
->addOption('ultron', 'Ultron')
->addOption('doctor-doom', 'Doctor Doom')
->setAtts(array('class' => 'marvelClass', 'data-mcu' => 'Marvel Villains'));
$mcu2->setContent($optGroup1.$optGroup2);
echo $mcu2;
```
Renders:
``` html
<select name="marvel_characters">
<optgroup label="Heroes" class="marvelClass" data-mcu="Marvel Heroes">
<option value="spiderman" dir="auto">Spiderman</option>
<option value="iron-man" selected dir="auto">Iron Man</option>
<option value="thor" dir="auto">Thor</option>
<option value="professor-x" dir="auto">Professor X</option>
<option value="black-widow" dir="auto">Black Widow</option>
</optgroup>
<optgroup label="Villains" class="marvelClass" data-mcu="Marvel Villains">
<option value="loki" dir="auto">Loki</option>
<option value="magneto" dir="auto">Magneto</option>
<option value="ultron" dir="auto">Ultron</option>
<option value="doctor-doom" dir="auto">Doctor Doom</option>
</optgroup>
</select>
```
I'm deliberating over creating a 'shortcut' method to add options without needing to specify the `value`, since it can either be derived from the label (dumbed down) or used directly in most cases. Not sure how best to name things, so it might not appear. But in the meantime, this optgroup capability is fairly handy.

**Comment by philwareham** [comment]
Minor thing, but `Textbox` seems close to `textarea` to me, but actually renders an `input`. I'd rather see `Input` and `Textarea` as the names.

**Comment by Bloke** [comment]
> I'd rather see Input and Textarea as the names.

Makes sense. Since everything's an 'input' I was just trying to be more specific. Let's review the names and come up with a table of what we think makes sense for the various input controls.

**Comment by Bloke** [comment]
Okay, that's a pretty good suggestion. See first post where I've just put the current implementation list of helper functions. If you want to add the proposals there we can see what makes sense all in one place.

Essentially, every tag is constructed from either the `Tag` or `TagCollection` so if you want to make bare bones tags or customise anything to the Nth degree, go there. Otherwise, use the 'helper' methods that bundle up the Tag/TagCollection calls into convenient-to-swallow bundles that cover the most useful options. Just saves typing, that's all.

**Comment by Bloke** [comment]
Renamed `Textbox` and `TextboxSet` as `Input` and `InputSet`, respectively.

Is everyone okay with calling this a 'Widget' library? I just chose the name as that's what I call them - input widgets - but if it's better in industry parlance to call them something else, then let's rename the class now.

Maybe `UI`? Then we'd use:

```
Txp::get('\Textpattern\UI\Input', 'name', 'value');
```

Is that better? Anything else we could/should use instead?

**Comment by Bloke** [comment]
For dates and times, I'm torn. Could do with some guidance.

We currently have `tsi()` which renders a single part of a date - year, month, day, hour, minute or second - as a text input box with a `inputmode=numeric`, a validation `pattern` plus `size`, `maxlength` and `placeholder` set.

That's all good, would be quite handy for us to have the same type of features in the widget library, but isn't very universal. What about people (plugin authors) who might want `<input type="date">` or `type="time"` or `type ="datetime"`? Should we cater for these as shortcuts too?

Dates and times can be rendered easily enough with what we have already:

```
$datebox = Txp::get('\Textpattern\Widget\Input', 'posted')
->setAtts(array('type' => 'date'))';
```
so do we really need convenience wrappers? e.g. 
```
Txp::get('\Textpattern\Widget\Date');
Txp::get('\Textpattern\Widget\Time');
Txp::get('\Textpattern\Widget\DateTime');
```

Or should we just stick to the approach we have now in `tsi()` and make convenience wrappers for individual date/time components? Maybe a suite of them:
```
Txp::get('\Textpattern\Widget\DateYear');
Txp::get('\Textpattern\Widget\DateMonth');
Txp::get('\Textpattern\Widget\DateDay');
Txp::get('\Textpattern\Widget\TimeHour');
Txp::get('\Textpattern\Widget\TimeMinute');
Txp::get('\Textpattern\Widget\TimeSecond');
```
That feels bloaty. So maybe a single one per 'type' that you can use a second argument to select which you want:

```
Txp::get('\Textpattern\Widget\InputDate', 'year');
Txp::get('\Textpattern\Widget\InputDate', 'month');
Txp::get('\Textpattern\Widget\InputTime', 'hour');
Txp::get('\Textpattern\Widget\InputTime', 'minute');
...
```
And maybe a convenience wrapper that amalgamates the individuals into combinations we use most in core to render the three input boxes as a complete set with separator spans between them.

Either would allow us to continue our tradition of separating out the date inputs into boxes, but it just feels kind of, I dunno, wrong to presume that's what core and plugin authors might use to render date and time widgets now or in future.

Maybe we don't need dedicated date/time input classes and we just leave it to people to design their own? We don't do it a lot in core and there are so many ways people could wish to render dates (e.g. months as names, localised names, or numbers) that we can't hope to provide widget shortcuts for all of them. But the library feels complete without _some_ form of date/time rendering. The question is... how far do we go?

Thoughts?

**Comment by philwareham** [comment]
I’d go with UI as the terminology. I’ll read through the rest of god messages when back from holiday. Cheers Stef.

**Comment by Bloke** [comment]
I'm inclined to agree about using `UI`. Just gotta decide what to call the methods when we 'add' other UI elements to it. e.g. when building up a `<form>` tag you might create a new `Txp::get('\Textpattern\Widget\Form', 'myPanel')` and then add an Input box, a few Checkboxes, a Select list to it. Currently we have `->addWidget()` so an alternative would be required:

`->addUI()` doesn't sound right to me.
`->addElement()` maybe?
`->addComponent()` too wordy?
`->addItem()` too vague?
`->addInput()` too confusing since we have an element called `Input`?

Anything jump out as being a good fit? Anything else we might consider?

**Comment by bloatware** [comment]
I'm fine with just `->add()`, unless we might need to add something else than form elements.

**Comment by Bloke** [comment]
Related: intending to add `<table>` support to the library.

As we migrate core over to this library, my intent is to:

1. replace all the code inside the functions in `txplib_forms.php` with calls to the new widgets, and deprecate all functions in this file.
1. probably replace most of the utility HTMLish code in`txplib_html.php` too and deprecate those functions. Bonus marks if we can get rid of them all.

With that in mind, we have a few functions that I don't see used:

* `fetch_editable()` seems to be unused in core. It seems only ever intended to be a private function (albeit unenforced) so I'm tempted to just axe it unless anyone can think of a reason to keep it.
* `type_options()` seems to be used once from the function `type_select()` inside `txplib_forms.php` itself, which is never called from anywhere else. Again, both seem to be implied private so ditch them?
* `timezoneSelectInput` was added in 4.7.0 but is used only once in the Prefs panel. Is it worth making a new timezone widget in the library and deprecating this function? Or shall we just remove it? We could build the timezone widget directly from the Prefs panel using regular Select/Option/Optgroup objects in this library. Just depends if we think there's value in other people creating a standard timezone dropdown.

**Comment by Bloke** [comment]
> I'm fine with just `->add()`

That's fine. Better and shorter. Good thinking. We currently have:

1. `->addWidget()` could be served by `->add()`
1. `->removeWidget()` could be served by `->remove()`
1. `->getWidget()` could be served by `->get()`

There's very little other 'add' functionality as far as I can recall (we `setAtts()`, `setTag()`, 'setAction()' and `setValue()` for example). We do `addOption()` to select list and optgroups, and can `addOptGroup()`.

Likewise there's nothing else we `remove`. There is, however, `getKey()` in a lot of classes, and `getIterator()` for collections. So having a `get()` on its own might be slightly confusing. I think that's why I suffixed it with `Widget` in the first place to make it clearer. But if we're dropping that 'widget' terminology, then we could just drop the suffix and hope it's still clear what we're adding, removing and getting.

**Comment by Bloke** [comment]
> I'm fine with just `->add()`

Alternatively, if we feel this might be a little terse/confusing, how about

1. `->addField()`
1. `->removeField()`
1. `->getField()`

??

**Comment by bloatware** [comment]
I thought of `addField()`, but was perplex about `addField('fieldset')`. Edge case, of course.

**Comment by Bloke** [comment]
> I thought of `addField()`, but was perplex about `addField('fieldset')`. Edge case, of course.

Uhhh yeah. That is awkward. Okay, scrap 'field'. Ummmm...

**Comment by Bloke** [comment]
Okay, in 2a33d598 I've moved everything to the `UI` namespace so any examples referenced in this issue using `\Textpattern\Widget` should use `\Textpattern\UI\ instead. Also, calls to `addWidget()` are now just `add()`. Let's see how we get on with this as we add the remaining few components to make this library fully rounded.

**Comment by philwareham** [comment]
Should we add details/summary (disclosure) to the UI widgets too? Possibly alert pills and boxes, progress bars, etc. too? Basically cherry picking elements out of here: https://hive-framework.philwareham.co.uk/

**Comment by Bloke** [comment]
Sure. Anything that'll save us coding common elements over and over when we use them in core should be represented.

Things like alert boxes... probably not, as we only use them once per theme and each one is different. But there are a bunch of elements there which do have widely applicable use cases and we use them throughout.

Disclosures, definitely. I'll see if I can add that today.

**Comment by philwareham** [comment]
Actually it’s best to use this as a guide https://design-patterns.textpattern.com/docs/

**Comment by Bloke** [comment]
So we're sticking with `<section><h3>heading</h3><div>content</div></section>` for now?

If so, it'll be a fab test of the power of this library because, in theory at least, if we move over to details/summary one day, it should only require changing the Disclosure object's internal rendering and structure - no change to any other code in core. Encapsulation at work :)

**Comment by philwareham** [comment]
Yep, details/summary not universally supported yet but that’s the future plan.

**Comment by Bloke** [comment]
Cool. Just so I can think about this, will details/summary require an id of the `<details>`container still? And an id for the anchor? Because those are the key bits of info I was planning to make the Disclosure object take as arguments when you make one. But if the IDs are not going to be required in future, it makes more sense to accept the summary text and content as arguments, then have the IDs added afterwards using `setAtts()`.

Presumably we'll still require some kind of pane ID so we can store the state of the twisty in localStorage? Statefullness doesn't seem to be part of the details/summary spec. Initial open/close of the widget goes on the outer element not the inner element but as long as we accept an open/closed property in the object, transitioning from one to the other is easy.

It probably makes more sense to take the pane ID (a.k.a. localStorage key) as an argument, and then infer the anchor ID from that? I know our current implementation often differs the name of the pane and the name of the anchor, but I can't see a reason not to make them the same and save a parameter. Do you think anybody targets the `name-of-details` directly in JavaScript? Do we in core?

**Comment by Bloke** [comment]
First pass at Disclosures has been done.

``` php
$content = Txp::get('\Textpattern\UI\Tag', 'p')
->setContent('First paragraph goes here.');
$content2 = Txp::get('\Textpattern\UI\Tag', 'p')
->setContent('Second paragraph goes here.');
$twisty = Txp::get('\Textpattern\UI\Disclosure', 'regionname1_group', 'panel-name')
->setLabel('I am the clickable heading')
->add($content)
->setHelp('some_help_topic_name')
->add($content2);
dmp($twisty->render());
echo $twisty;
```

Renders:

``` html
<section class="txp-details" id="regionname1_group" aria-labelledby="regionname1_group-label">
<h3 class="txp-summary" id="regionname1_group-label"><a href="#panel-name" role="button" data-txp-token="51670132b3c8679dcc7b695981dbe4f6" data-txp-pane="panel-name">I am the clickable heading</a>&#160;<a rel="help" title="help" role="button" class="pophelp" href="#"><span class="ui-icon ui-icon-help">help</span></a></h3>
<div class="toggle" id="panel-name" role="group" style="display:none">
<p>First paragraph goes here.</p>
<p>Second paragraph goes here.</p>
</div>
</section>
```
The arguments to the Disclosure constructor are:
* `id` to use as the surrounding wrapper (a `<section>` tag in this case, but that may change in future).
* `pane` to set the id of the localStorage id/reference for storage of the twistable state. If this is omitted, the twisty is not, well, twisty. It'll just be a static heading+content wrapped in a section marker.

You can use the `setVisible()` method to set the initial display of the twisty open (`true`) or closed (`false`, by default). Calling the method without any arguments will toggle the status, which is useful for Ajax calls.

The content can also be built using a `TagCollection` if you prefer so only one `->add()` is required.

Not sure if this is the best implementation yet, so improvements are welcome.

**Comment by philwareham** [comment]
OK, looking good Stef. Sorry I'm very time-limited right now but I guess the next step is to build out a panel using this library - as a proof of concept. Then we can look at the HTML code it produces and see where any improvements need to take place.

Realistically I'm going to be tied up with current deadlines until mid-October.

**Comment by Bloke** [comment]
No worries. I've got more elements to add yet. The big ones are file upload and inputLabel. Then maybe table support somehow. And perhaps some low-level stuff as convenience objects to replace `graf()` and `span()` and `href()` if we think it's warranted.

Then, yeah, I think rolling it into dev and rewriting a single panel as a test will highlight places we can make optimisations or introduce new objects to make the process of building a UI simpler.

**Comment by philwareham** [comment]
@bloke I guess now (or at least after any critical 4.8.1 work) is a good time to revisit this UI library and start actually using it to build admin panels? Just keeping it on the radar.

**Comment by Bloke** [comment]
Absolutely. There's two more things to do that I consider essential (both of which I was partway through and need to finish) :

* File upload form
* Input + Label

After that, it'd be really nice to put table stuff in there - to replace `tr()`, `td()`, `tdcs()`, `tdrs()`, `column_head()`, `column_multi_head()` the comically titled `assHead()` etc but I think that's going to need some packaging so it can be added afterwards.

**Comment by Bloke** [comment]
Okay, finally, it's in dev. See 310643a3.

Still got Table support to go - that's in progress - and maybe a few straggler convenience functions. Perhaps add a PopHelp class in the UI library and instantiate it from `Module/Help/HelpAdmin->pophelp()` instead of having the rendering occur there.

In terms of what we can eventually jettison from txplib_html.php and txplib_forms.php, most of it can be replaced by the UI classes. We can go through and deprecate the function content there as we grow more comfortable with the UI library, then perhaps swap out the functionality with calls to the library so they remain as mere stubs. That'll cut down file sizes a tad to offset the larger UI library size.

There are a few convenience functions in txplib_forms.php that we don't even use and I'm not sure are needed, but will take advice:

### Not used
* `small()`
* `comment()`
* `htmlPre()`
* `fetch_editable()`
* `type_select()`
* `type_options()`

### Used
* `strong()`
* `span()`
* `href()`
* `timezoneSelectInput`
* `tsi()`

I'm tempted to let the unused ones lapse. Of the used set, I think we might introduce a UI library class to replace the date/time functions, as they're potentially useful bundled functionality. But the others, I'm not sure about. They don't add much, but could add a little clarity/shortening of code so may be worth considering.

By way of comparison, we're talking this syntax today:

```
Txp::get('\Textpattern\UI\Tag', 'strong')
   ->setContent('stuff to bold');

Txp::get('\Textpattern\UI\Tag', 'span')
   ->setContent('stuff to span');

Txp::get('\Textpattern\UI\Tag', 'a')
   ->setContent('My link')
   ->setAtt('href', 'https://...');
```

vs this if we were to introduce wrappers:

```
Txp::get('\Textpattern\UI\Strong', 'stuff to bold');
Txp::get('\Textpattern\UI\Span', 'stuff to span');
Txp::get('\Textpattern\UI\Anchor', 'My link', 'https://');
```

Anchor is arguably useful. Might add that. Maybe the rest too, for code shortening. What do you think?

Note that syntax can be shortened further by using a `use` statement at the top of any files that require UI stuff, AND using objects instead of `Txp::get()`:

```
use \Textpattern\UI\Tag;
...
$strong = new Tag('strong');
$strong->setContent('stuff to bold');

$span = new Tag('span');
$span->setContent('stuff to span');

$href = new Tag('a');
$href->setAtts(array(
   'class' => 'button',
   'href' => 'https://...',
))
->setContent('My link');
```

vs:

```
use \Textpattern\UI\Strong;
use \Textpattern\UI\Span;
use \Textpattern\UI\Anchor;
...
$span = new Span('stuff to span');
$strong = new Strong('stuff to bold');
$href = new Anchor('My link', 'https://...');
$href->setAtt('class', 'button');
```

So yeah, maybe it is worth adding them for brevity.

Loads of other little wrappers we could add if we felt it necessary, but the big hitters (bar table) are in there now.

**Comment by Bloke** [comment]
Alright, Anchor and Span do have use. Added to the library in fcf46b6b and 89ba1a12.

### Anchor usage
```
$anchor = Txp::get('\Textpattern\UI\Anchor', 'Ajax link here');

$anchor2 = Txp::get('\Textpattern\UI\Anchor', 'Parameter link here',
    array(
        'arg1' => 'bunch',
        'arg2' => 'of',
        'arg3' => 'params',
    )
); 

$anchor3 = Txp::get('\Textpattern\UI\Anchor', 'Link with attributes', 'https://textpattern.com/')
->setAtts(array(
    'data-set' => 'some value',
    'id' => 'some_id',
    'class' => 'button',
)); 

echo $anchor;
echo $anchor2;
echo $anchor3;
```

Renders:
```
<a href="#">Ajax link here</a>
<a href="?arg1=bunch&amp;arg2=of&amp;arg3=params">Parameter link here</a>
<a href="https://textpattern.com/" data-set="some value" id="some_id" class="button">Link with attributes</a>
```

### Span usage
```
$span = Txp::get('\Textpattern\UI\Span', 'Content of my span'); 
$span2 = Txp::get('\Textpattern\UI\Span', 'Span with atts')
->setAtts(array(
    'data-set' => 'some value',
    'id' => 'some_id',
    'class' => 'some-class',
)); 
echo $span.br;
echo $span2.br;
```

Renders:
```
<span>Content of my span</span>
<span data-set="some value" id="some_id" class="some-class">Span with atts</span>
```

**Comment by Bloke** [comment]
I've added AdminAction (for sLink, eLink, aLink and wLink) and AdminDelete (for dLink) so we can create (for example) sLink() type functionality this way now:

```
$slink = Txp::get('\Textpattern\UI\AdminAnchor', 'section', 'section_edit', gTxt('create_section'));
```
Which renders:
```
<a href="?event=section&amp;step=section_edit">Create section</a>
```

If you want to get fancy:

```
$seclink = Txp::get('\Textpattern\UI\AdminAnchor', 'section', 'section_edit', gTxt('create_section'))
->setProperty('token', true)
->setParam('id', 'some-id-value')
->setParams(compact('page', 'search_method', 'crit', 'dir', 'sort'))
->setAtt('class', 'txp-button');
```
Which renders something like:
```
<a class="txp-button" href="?event=section&amp;step=section_edit&amp;id=some-id-value&amp;page=2&amp;search_method=name%2Ctitle&amp;crit=txp&amp;dir=asc&amp;sort=name&amp;_txp_token=fbd562...">Create section</a>
```

You can also add a 4th parameter to the constructor: 'get' (which it is by default, to render an anchor) or 'post' (to render a form). If you do that, the 'linktext' (3rd param) should be a proper Input element like a `<button>` so it can be used to submit the form, instead of a simple bit of text.

The AdminDelete class is simply a shortcut for dLink() which sets up suitable defaults for rendering 'x' buttons and the like. Very similar to AdminAction:

```
$dlink = Txp::get('\Textpattern\UI\AdminDelete', 'image', 'thumbnail_delete', 'post')
->setParam('id', 'the-id-to-remove')
->setParams(compact('page', 'search_method', 'crit', 'dir', 'sort'));
```

Renders:
```
<form data-verify="Really delete?" method="post" action="index.php">
<button class="destroy" type="submit" title="Delete" aria-label="Delete"><span class="ui-icon ui-icon-close">Delete</span></button>
<input name="event" type="hidden" value="image">
<input name="step" type="hidden" value="thumbnail_delete">
<input name="id" type="hidden" value="the-id-to-remove">
<input name="page" type="hidden" value="2">
<input name="search_method" type="hidden" value="name,title">
<input name="crit" type="hidden" value="txp">
<input name="dir" type="hidden" value="asc">
<input name="sort" type="hidden" value="name">
<input name="_txp_token" type="hidden" value="cad847...">
</form>
```

You can use `->setProperty('verify', 'some_custom_string') to customise the verify message (it'll be passed through `gTxt()` for you);

**Comment by Bloke** [comment]
Next question then:

We currently have `graf()` (shorthand, I presume, for paragraph) which renders 'p' tags and associated attributes.

If we were to represent this in the UI library, what should it be called, bearing in mind all the other classes have reasonable, full names? Ideas:

* Paragraph (complete, unambiguous, a bit wordy).
* Graf (shorthand, quirky, kind of backwards-compatible name to what we have, but doesn't fit with the rest of the library).
* P (super short, perhaps not very readable in code).
* ... something else?

Any thoughts on this, @philwareham @bloatware @petecooper or anyone?

**Comment by philwareham** [comment]
`Para` could be an option?

**Comment by petecooper** [comment]
Above my pay grade, alas. Paging @jools-r for potential input.

**Comment by Bloke** [comment]
`Para` would work, thanks. Could go with that for the time being and if anyone comes up with anything better in the meantime we can switch it out.

**Comment by philwareham** [comment]
Are we enforcing any `class` names on the UI elements? For example, see the UI styling design patterns doc where certain names are required to style some of the UI elements: https://design-patterns.textpattern.com/docs/

I wonder if we can (temporarily) generate a sheet within Textpattern for this new UI that follows the same idea as the UI styling design patterns above, so we can see as many of the new UI widgets together on a single page and work out which need additional attributes added and suchlike?

If we do add core `class` names, I _think_ there would need ti be criteria for an author to:

a) add additional classnames
b) omit the core classname (maybe?)

**Comment by Bloke** [comment]
Yes. If there are any enforced classes (in the bigger classes such as Disclosures and UploadForm and InputLabel; even Checkbox) they're preserved from today's core. You can override the classes if you wish, and some options will remove or move them just like today's functions (e.g. if you remove wrappers). But there's not a tag-level way of _appending_ to a key-value pair yet. We could probably add that if it's useful, but it will apply to _all_ (non-boolean) attributes.

Here's an example of current functionality with InputLabel:

```
use \Textptattern\Input;
use \Textptattern\InputLabel;

$input = new Input('field-name');
$input
   ->setBool('autocomplete')
   ->setAtts(array(
      'data-extra' => 'attached-stuff',
      'class' => 'input-class'
   ));

$ilab = new InputLabel('field-name', $input, 'fieldLabel');
$ilab->setHelp('field-explainer');

echo $ilab;
```

That will render:
```
<div class="txp-form-field edit-field-name">
<div class="txp-form-field-label">
<label for="field-name">fieldLabel&#160;<a rel="help" title="Help" role="button" class="pophelp" href="?event=help&amp;step=pophelp&amp;item=field-explainer"><span class="ui-icon ui-icon-help">Help</span></a></label></div>
<div class="txp-form-field-value">
<input name="field-name" type="text" data-extra="attached-stuff" class="input-class" autocomplete>
</div>
</div>
```

If you override the class name at the InputLabel level, it overrides it in the wrapper entirely:

```
$ilab = new InputLabel('field-name', $input, 'fieldLabel');
$ilab
   ->setHelp('field-explainer')
   ->setAtt('class', 'override');
```

You get:
```
<div class="override">
<div class="txp-form-field-label">
<label  ...
...
</div>
</div>
```

If you remove the wraptag from the InputLabel:

```
$ilab = new InputLabel('field-name', $input, 'fieldLabel');
$ilab
   ->setHelp('field-explainer')
   ->setAtt('class', 'override')
   ->setWrap('');
```

Then you see a much slimmer result:

```
<div class="override">
<label for="field-name">fieldLabel&#160;<a rel="help" title="Help" role="button" class="pophelp" href="?event=help&amp;step=pophelp&amp;item=field-explainer"><span class="ui-icon ui-icon-help">Help</span></a></label>
<input name="field-name" type="text" data-extra="attached-stuff" class="input-class" autocomplete>
</div>
```

There's currently no (neat) way to get at the wraptag classes (`txp-form-field-label` and `txp-form-field-value`) as I didn't see it necessary. We could expose those if you felt it warranted.

I've got an article on my dev site with all the new tags in use and various attributes applied as a testing ground but yes I'd like to extend this to an official list so we can check we've covered everything. And also to make sure we've captured as many default use cases as we can so there's less repetition required in core and plugins when using these things for standard cases.

**Comment by philwareham** [comment]
Sounding great, thanks Stef!

> But there's not a tag-level way of appending to a key-value pair yet. We could probably add that if it's useful, but it will apply to all (non-boolean) attributes.

If we can get a clean way of appending, that would be good. But we need somehow to allow append **or** override for key-value pair, which I'm unsure how we can do in an obvious way. I guess the other option would be for authors to add the standard core class in their override such as...

    ->setAtt('class', 'txp-button myclass')

...but that assumes the author is fully aware of core class names.

> There's currently no (neat) way to get at the wraptag classes

That seems fine, don't see a need to change those - authors can alway target a specific `txp-form-field-label` and `txp-form-field-value` element via a `class` on the parent `div`.

**Comment by Bloke** [comment]
Well you can cheat in _some_ of the classes, such as Checkbox:

```
$cb = new \Textpattern\UI\Checkbox('myCB', 'fruit', false);
$cb
   ->setAtts(array('data-special' => 'some-custom-thing', 'id' => 'custom-id'))
   ->setAtt('class', $cb->getAtt('class') . ' extra-class');
```

That renders this where the extra-class is appended to whatever we fetched via `getAtt()`:
```
<input class="checkbox extra-class" id="custom-id" name="myCB" type="checkbox" data-special="some-custom-thing" value="fruit">
```

Sadly, we can't always do that. InputLabel, for example, doesn't create any of the internal classes it uses until it comes time to output the tag (in the `render()` method) so you can't access such built-in properties of the class until it's too late. If we could somehow pre-load the classes that the wrapper was going to use, then it would be possible to append that way. I might see if it's possible to fudge things like that in as many of the classes as possible: 1) for consistency, 2) for flexibility., 3) so we can introduce an `appendAtt()` method which does the "`getAtt()` and tack on the rest" trick internally to save you doing it by hand (which has the side benefit that you can use `Txp::get()` instead of being forced to use the `new` syntax so you have a handle to the object.

Incidentally, `getAtt()` takes an optional second parameter: the default value to return if that key is null.

**Comment by philwareham** [comment]
OK, I think that covers all eventualities. If we can get a dummy page together that renders as many of these things as possible, we can see how it all works, how it is styled, and what more needs to be added (if anything).

I can always write more CSS for any additional UI widgets we see could be beneficial for future layouts and UI additions.

**Comment by Bloke** [comment]
Commit 04a50ecb introduces `appendAtt()` and `appendAtts()`. Example:

```
$cb = Txp::get('\Textpattern\UI\Checkbox', 'myCB', 'fruit', false)
->setAtts(array(
   'data-special' => 'some-custom-thing',
   'id' => 'custom-id'
))
->appendAtt('class' , 'extra-class')
->appendAtts(array(
   'class' => 'more-class-goodness',
   'data-onload' => 'some_script',
));

echo $cb;
```
That renders:
```
<input class="checkbox extra-class more-class-goodness" id="custom-id" name="myCB" type="checkbox" data-special="some-custom-thing" data-onload="some_script" value="fruit">
```

Note that if the thing you're appending doesn't exist, it's created.

Both methods accept a final parameter: `glue` which is a string to join your new value(s) with any existing ones. It defaults to a single space. Might be useful if you're appending stuff to comma-separated lists of stuff in an attribute.

**Comment by Bloke** [comment]
Shall we close this since it's now merged, or leave it open for the time being in case we need to revisit it as we start to slowly introduce the classes into the core panels?

**Comment by philwareham** [comment]
Have we addressed https://github.com/textpattern/textpattern/issues/1247 and https://github.com/textpattern/textpattern/issues/1235 too? If so then yes we can close - otherwise I’d like to include those issues in the widget work.

We probably need to then open a new issue for converting panels to use new ui.

**Comment by Bloke** [comment]
We're still using the old disclosure style. Easy to switch so I can tackle #1247 if support is wide enough for details/summary.

For the checkboxes, do we still just need to [add the class you noted](https://github.com/textpattern/textpattern/issues/1235#issuecomment-380032823) by default so you can style the rest using Adrian Roselli's write-up as a guide? Seems that's pretty comprehensive, including keyboard use, high contrast mode, dark mode and RTL.

**Comment by philwareham** [comment]
> We're still using the old disclosure style...

We have dropped IE11 support in 4.9 onwards so all the browsers we support can do details/summary disclosures. Hopefully the code should be easier because we shouldn't need to bootstrap ARIA onto the markup as disclosures already have accessibility built-in as default.

> For the checkboxes, do we still just need to add the class you noted...

@bloke, I think if we put a `class` of `txp-switch` on the parent `div` of radio button and checkbox elements, that should make those areas targetable to CSS and skinnable.

So the `div` would be:

```
<div class="txp-form-field-value txp-switch">
    <input class="radio active" id="example1" type="radio" checked>
    <label for="example1">No</label>
    <input class="radio" id="example2" type="radio">
    <label for="example1">Yes</label>
</div>
```

**Comment by philwareham** [comment]
The other concern I have with using switches is that common UX expectations for them these days is that they instantly save a pref when switched. Our prefs panel doesn't do that (you have to then 'save' the prefs). By using a switch we risk user confusion unless you start AJAX-saving the prefs - which may be undesirable.

**Comment by Bloke** [comment]
That is a valid point re: immediacy.

If we go the Ajax save route then it gives nobody a way of backing out changes. And it'd have to work like that for all prefs - including the admin theme switcher and so forth.

That also means no Save button, which potentially renders a bunch of plugins that utilise the prefs panel useless, as they won't necessarily be wired for Ajax saves (it requires a change of type for Type 3 plugins to Type 4, at minimum).

**Comment by Bloke** [comment]
In terms of code it does mean some other weirdness. Anything checkboxy cascades from the UI\Checkbox class. Anything Radioish cascades from the UI\Radio class. So if we add the `txp-switch` class to those by default, then _all_ radios (including three/four/more choices) and checkbox sets (not that we use them yet in core) would get these classes.

If we move the enforced `txp-switch` class to just those classes that offer 'binary' choices - UI\YesNoRadio and UI\OnOffRadio - then if you decide to use the bog standard UI\RadioSet and pass it only two options, you _won't_ get the `txp-switch` by default. Same with checkboxes. That's inconsistent, because the difference between a checkbox toggle and a two-choice radio is usually just down to labelling.

The simplest solution here is to simply add the styles to the admin themes' CSS for handling them, then add the class when we build the controls on an as-needed basis. So the UI library stays pure and we just manually `->setAtt('class', 'txp-form-field-value txp-switch')` whenever we construct toggles we know need to behave this way. That's more hassle/code for us but is way more flexible.

**Comment by philwareham** [comment]
> So the UI library stays pure and we just manually...

OK, that works for me - so I can do the (probably optional) CSS for this, how would the following current HTML be changed by your recommendation (feel free to edit the code below)?:

```
<div class="txp-form-field-value">
    <input class="radio active" id="example1" type="radio" checked>
    <label for="example1">No</label>
    <input class="radio" id="example2" type="radio">
    <label for="example1">Yes</label>
</div>
```

**Comment by Bloke** [comment]
One thing we don't have in core right now but could really do with are numeric inputs and range sliders. These bring some unique challenges to a UI library because they have some optional components:

* min
* max
* step

The library itself can of course manage these if you build a UI\Input control by hand and pass in the relevant params:

``` php
use \Textptattern\Input;

$nb = new Input('rating');
$nb->setAtts(array(
    'type' => 'number',
    'min' => 1,
    'max' => 10
));

echo $nb;
```

That would output:

``` html
<input name="rating" type="number" min="1" max="10">
```

But there are some interesting caveats here. Namely we can't use '0' as a value without overriding the default flag behaviour to 'strip empty':

``` php
$nb = new Input('howmuch');
$nb->setAtts(array(
    'type' => 'number',
    'min' => 0,
    'max' => 1,
    'step' => '0.1'
), array(
    'flag' => TEXTPATTERN_STRIP_NONE,
));
echo $nb;
```

The same thing applies to range sliders.

Should we introduce a Number class that can handle these in a simpler fashion that would automatically apply the correct stripping flag? e.g.

``` php
$nb = new Textpattern\UI\Number('howmuch', 0, 10, 2);
```
(with the optional three values being min, max, step).

I kinda like that, but maybe someone has a better idea? It might not be worth going as far as creating a Range class, since it's identical under the hood to a Number: it just changes the `type` which is easy enough to do with a single attribute. Although we could add one for completeness as it's easy to extend the Number class with a single line of code in a dedicated Range class.

From a usage standpoint, we have another issue with these input types: where to store the sizing info. Let's consider prefs. If we want to expose a numerical input field for, let's say SMTP Port, the value can be (ostensibly) from 1 to 65535. So it makes sense to have a number input with min=1 and max=65535. Same could be said for any numeric inputs we have that are currently using text_input (logs_expire_days, max_url_length, etc).

The problem is that in the txp_prefs table we have no way of specifying these constraints. We have columns for the pref name, its value, its group ('event') and input type (`html`), visibility, position and scope (private/global).

We could really do with some additional database fields for specifying things like min, max, step, pattern, size and so forth. But not all of them are applicable to all field types so we'll probably need to make some design/storage compromises.

One obvious solution is to introduce a new `size` column. This means that for text_input types we can specify the small/medium/large/xlarge sizes directly in the pref value if we wish, instead of having to hard-code if we want to deviate from 'medium' in the panel code. I like that. We could perhaps reuse that field for numeric prefs too, e.g. allow a comma-separated list of sizes: min, max, step. Any missing values get defaults.

Not sure what 'size' would mean for YesNoRadio or OnOffRadio prefs. Guess we just have to take the hit on the wasted storage space. Multi-select lists could use 'size' to dictate how many items to display in the list before a scrollbar is shown, perhaps. Textareas could specify `rows, cols` - is that something we need or does CSS offer everything here?

Do we go further and introduce a `pattern` column too? That's really only applicable to freeform entry fields so would be wasted for numeric prefs and select lists and, well, pretty much any other type. But that would allow us to offer in-browser validation and also to dynamically build a PrefConstraint when prefs are saved, which would validate the value on the server-side before storage using the same pattern.

That'd be a boon for us (and links to issue #1600 where we wish to validate prefs before saving them to avoid bad values being stored). And it's a boon for plugin authors who want to store prefs as they can employ the same validation rules and have core do the hard work of enforcing their rules by comparing stored values to the pattern and kicking up a fuss if it doesn't match.

I'm just not sure if introducing a column specifically for 'pattern' is too restrictive or a waste of DB storage space. For numeric types, the min/max/step _is_ actually a storage constraint as well as a pattern-matching constraint AND a browser validation constraint, even though the actual HTML input controls don't support `pattern` explicitly for these types of input.

Since we don't really need both `size` and `pattern` for most types, can we just introduce a single additional column in txp_prefs that can cover both? Maybe call it `constraint` and then whatever is given in that column is dependent on the input control. By default, Radio buttons would validate to 0 or 1 unless you overrode it (somehow). Numbers validate to `min, max, step`. Input controls could validate to an arbitrary pattern constraint, but we'd have no way of specifying the input control size (width), short of adding an attribute manually when it's built, which means hard-coding it in the prefs panel code.

Alternatively - and very very dirtily - we could let the `constraint` column accept some key=>value or JSON data that specifies exactly which things we want to validate. e.g. for the image_dir pref, the DB fields could contain:

```
val: ""
type: 0
event: admin
html: text_input
position: 40
constraint: {
   "size": "medium",
   "pattern": "[a-zA-Z0-9\-\_]"
}
```

That's not very clean and purists would run a mile at the abuse of data storage "one field = one value" conventions, but it does offer us superior flexibility to handle any and all size/validation opportunities on both browser and server sides in a single compact field.

I'm torn. What do y'all think? Paging @bloatware @philwareham @petecooper @jools-r for comment.

**Comment by petecooper** [comment]
I'm afraid I won't be much help with this, given the paged audience I will respectfully defer to others.

**Comment by Bloke** [comment]
No probs. Thanks for looking anyway.

It's one of those conundrums of do we do it "right" (which involves a lot of tables and joins and slowdown) or do we do it "wrong" (which is way more flexible but yucky and less intuitive for plugin authors) or do we leave it in code (which is harder to maintain and not as useful to us or plugin authors) or some hybrid of the above/something else.

None of them are attractive, but since we're building a UI library that is, by its nature, going to be the cornerstone of building our interfaces, it pays to do it in the best way for the project to make the leanest code and least hassle for people using it, regardless of rightness and wrongness.

**Comment by philwareham** [comment]
I’m for the way that benefits us best, so if two key values (or more) were in the constraints field and that gave us maximum flexibility without hard coding then that seems ok to me.

**Comment by Bloke** [comment]
Yeah, I'm leaning that way as it does mean we can extend this in really neat ways in future to help our UI features. I just fear the *rolleyes* from the 4NF crowd who stare agape at our database structure and storm the castle with searing pitchforks for crimes against database design.

**Comment by bloatware** [comment]
I'd report it here: setting a cf render to `checkboxSet` (in `custom-fields` branch) and trying to access `Write` tab yields
```
 Uncaught Error: Call to undefined method Textpattern\UI\CheckboxSet::setAtt() in C:\xampp\htdocs\textpattern\textpattern\vendors\Textpattern\Meta\Field.php:862
```

**Comment by bloatware** [comment]
Another issue: `<input type="datetime" />` used in `dateTime` render is obsoleted in favour of `<input type="datetime-local" />`.

**Comment by Bloke** [comment]
Fixed the first one. Note to self: TagCollections don't have directly addressable IDs

**Comment by Bloke** [comment]
> Another issue: `<input type="datetime" />` used in `dateTime` render is obsoleted in favour of `<input type="datetime-local" />`.

Surely this can just be changed without any b/c issues anywhere we render such a widget?

**Comment by bloatware** [comment]
Nope, it still outputs `type="datetime"` for me. This value has been [dropped](https://github.com/whatwg/html/issues/336) about 8 years ago, so we should be safe replacing it with `datetime-local`.

**Comment by Bloke** [comment]
Ah balls yeah, the UI library short circuits it by using `strtolower()`. Hang on.

**Comment by Bloke** [comment]
Hmm, not very clean but since dateTime is what we refer to a date-time widget internally, I've [fudged it](https://github.com/textpattern/textpattern/commit/d6d0c8ea373b16ad6ba4ef8fca48c3d8f4b4dacb) in the renderer. Is that any closer?

**Comment by bloatware** [comment]
This has nailed it, thank you. Should we also switch to `dateTime` render in core datetime inputs (posted, expires etc) for consistency?

**Comment by bloatware** [comment]
Sorry @Bloke, when creating a date/time cf I get this:
```
Fatal error: Uncaught mysqli_sql_exception: Table 'txpdev.txp_meta_value_datetime-local' doesn't exist in C:\xampp\htdocs\textpattern\textpattern\lib\txplib_db.php:442
```

**Comment by phiw13** [comment]
PHP 8.2.11 / MySQL 8, on a fresh CF-branch install
While trying to create a `dateTime` field:

```
Fatal error: Uncaught mysqli_sql_exception: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '-local DEFAULT NULL, UNIQUE KEY meta_content (me' at line 5 in /Users/username/Sites/_txp-dev/textpattern/lib/txplib_db.php:442 
Stack trace: 
#0 /Users/username/Sites/_txp-dev/textpattern/lib/txplib_db.php(442): mysqli_query(Object(mysqli), 'CREATE TABLE IF...', 0) 
#1 /Users/username/Sites/_txp-dev/textpattern/lib/txplib_db.php(864): safe_query('CREATE TABLE IF...', false) 
#2 /Users/username/Sites/_txp-dev/textpattern/vendors/Textpattern/Meta/Field.php(389): safe_create('txp_meta_value_...', 'meta_id int(12)...') 
#3 /Users/username/Sites/_txp-dev/textpattern/include/txp_meta.php(587): Textpattern\Meta\Field->save(Array) 
#4 /Users/username/Sites/_txp-dev/textpattern/include/txp_meta.php(553): meta_save() 
#5 /Users/username/Sites/_txp-dev/textpattern/include/txp_meta.php(70): meta_save_ui() 
#6 /Users/username/Sites/_txp-dev/textpattern/index.php(238): include('/Users/username/S...') 
#7 {main} thrown in /Users/username/Sites/_txp-dev/textpattern/lib/txplib_db.php on line 442
```

–^–

And while playing, I noticed that several CF input fields have an invalid `size` attribute: input type = date | time | number | range (datetime-local is probably affected too).

**Comment by Bloke** [comment]
The latest slew of custom field commits should fix the above two errors. Apologies for the humongous delay. Not sure what you mean by an invalid `size` attribute. Things like Number can take an input field size (width), can't they?

**Comment by phiw13** [comment]
Nope…, or rather, not  really. `<input type=number size=32 value min=0 max=100>` is definitely not valid, try validating this test: https://dev.l-c-n.com/_temp/n-d.html (under the error message the validator helpfully lists the input types where `size` applies [*]). Curiously for `type=date` the validator does not complain :-( – anymore as it used to consider that invalid.

See [MDN type=number](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number.) It lists all valid attributes: type & value, min, max, step, placeholder, readonly. And “list” (+ global attributes). Down the page, it only notes that `size` has no effect. Ambiguity galore, and nowhere clearly defined.

same for date/time/date-time/ month/week

Bottom line: the `size attribute is only relevant for the old - HTML 4 – controls: [type=text/password], select[multiple]. If possible the attribute should not be applied – on the Edit File panel, “Download count” number field, the size attribute is correctly omitted (field lacks a `<label />` btw); idem ditto on the Preferences panel, default install.

[*] when type is text, search, url, tel, email, or password – by which they possibly mean, the attribute only has an effect on those input controls.

**Comment by Bloke** [comment]
Weird. I changed one of the input fields in the inspector to size=10 and it made it smaller. But whatevs. I'll see if I can remove it somehow. Thanks for the info.

**Comment by phiw13** [comment]
> I changed one of the input fields in the inspector to size=10

You tested that with Firefox and `input type=number`, I guess. Try with anything Chromium based or Safari 14~18. The size attribute has no effect, incl Firefox. Or try the `type=date` field in my test file above.

**Comment by Bloke** [comment]
Removed `size` on:

* date
* time
* dateTime
* number
* range

Incidentally, that seems to have also fixed some bizarre behaviour I had on datetime fields in Firefox where I could only click on any cell in the popup date picker once, whereby it would just drop out of the popup and I'd have to click again to change (e.g.) to the next month.
